### DISCUSSION TITLE
Simplified self type on operators - typeof(this), sizeof(this), nameof(this)

### DISCUSSION THREAD

**Discussion by Rekkonnect** [dotnet/csharplang#5261] [Language Ideas]
Straight to the point, the `this` qualifier is useless inside `typeof`, `sizeof`, and `nameof` operators. This proposal aims to reserve their usage for simplified referral to the current type.

```csharp
public class SomeType
{
    void M()
    {
        var thisType = typeof(SomeType);
        var thisName = nameof(SomeType);
    }
}
```
would become
```csharp
public class SomeType
{
    void M()
    {
        var thisType = typeof(this);
        var thisName = nameof(this);
    }
}
```

Since those operators are performed at compilation time, the runtime object type does not affect the result. In other words, usage of code with those operators on derived types will not change the result. Given the above example, assume
```csharp
public class OtherType : SomeType
```

`thisType` and `thisName` will be `typeof(SomeType)` and `nameof(SomeType)` respectively, disregarding `OtherType` as a candidate. This concerns consistency around the operators, and does not introduce imlpementation complexity, and resulting code cluttering.

**Comment by CyrusNajmabadi** [comment]
The `nameof` case is pretty strange as it violates the current aspect of `nameof` that it literally uses the identifier passed into it.  In other words, nameof doesn't care what is pointed at.   For example, if you have:

```c#
using System;
using X = System.Int32;

public class P
{
    public static void Main()
    {
        Console.WriteLine(nameof(X));
    }
}
```

Then you get `X` printed, not `Int32`.  This would violate that as it would not print `this` but would instead print `SomeType`.  

This definitely breaks my expectation as nameof is literally for getting the name value of that token as a string.  But there is no name there whatsoever.

**Comment by TahirAhmadov** [comment]
The `nameof(this)` issues notwithstanding, I think the `typeof(this)` and `sizeof(this)` should be possible and would definitely help. For me, the `typeof(this)` is very important, because a few classes have long names and a few generic arguments, and `typeof(this)` would go a long way to make the code more readable in those situations.
Also, they should work in methods/props/events, as well as field initializers, in both instance and static contexts.
